### DISCUSSION TITLE
Exclude path from matching in redirects/rewrites/headers

### DISCUSSION THREAD

**Discussion by balazsorban44** [vercel/next.js#24255] [Help] [answered]
Hi there!

I am looking at Next.js redirects https://nextjs.org/docs/api-reference/next.config.js/redirects and headers https://nextjs.org/docs/api-reference/next.config.js/headers.

I want to match ALL paths, EXCEPT some.

As far as I can tell, Next.js uses `path-to-regexp` to write rules for the source matching.
Looking at their repo, I found two related issues:
https://github.com/pillarjs/path-to-regexp/issues/158, https://github.com/pillarjs/path-to-regexp/issues/141

Both have been closed, and if I understand it correctly, you have to use a regular expression, instead of a string, which seem to not be supported in the `next.config.js` file.

Here are my attempts:

1. Here I want to add a security header, but allow iframing certain paths:

```js
// Headers
{
  source: "match every page, except if it begins with /foo or /bar",
  headers: [
    {
      key: 'X-Frame-Options',
      value: 'DENY,
    },
  ],
}
```

2. This one is using the new `has` field https://nextjs.org/docs/api-reference/next.config.js/rewrites#header-cookie-and-query-matching
```js
// Redirects
{
  source: "match every page, but exclude the ones that have /outdated in them",
  has: [
    {
      type: "header",
      key: "User-Agent",
      value: "(?<agent>MSIE)",
    },
  ],
  permanent: false,
  destination: "/outdated?agent=:agent",
},
```
Basically what I want is if the user has a user-agent indicating IE10 or older, they should be redirected to a page. (Yes, this is a valid case, unfortunately). But if the user is already on the outdated page, the path shouldn't match, otherwise, a redirect loop happens.

Any suggestions?

**Comment by ijjk** [comment, marked as answer]
Hi, you can achieve this with a negative look-ahead e.g.

```js
module.exports = {
  redirects() {
    return [
      {
        source: '/:path((?!_next|foo$|bar$).*)',
        destination: '/another',
        permanent: false
      }
    ]
  }
}
```

Note: `_next` is included to prevent redirecting Next.js' required static files.

**Reply by balazsorban44** [reply]
I now verified, and this works as expected! Thank you so much! :heart: 

My question is, how does that work with ISR? Won't this be cached in a CDN? Will that be bypassed somehow? Is it up to the CDN? Could you provide some starting points on this?

I leave my code here for future reference:

```js
// next.config.js
module.exports = {
  async redirects() {
    const cmsRedirects = await getRedirects()
    return [
      {
        source: "/((?!_next|outdated).*)",
        has: [
          {
            type: "header",
            key: "User-Agent",
           // http://useragentstring.com/pages/useragentstring.php?name=Internet+Explorer IE10<
            value: ".*MSIE.*",
          },
        ],
        permanent: false,
        destination: "/outdated",
      },
      ...cmsRedirects,
    ]
  },
  headers() {
    return [
      {
        source: "/((?!foo).*)",
        headers: [{ key: "X-Frame-Options", value: "DENY" }],
      },
    ]
  },

}
```

**Reply by ijjk** [reply]
The caching should be up to the CDN as long as `permanent: false` is used

**Reply by balazsorban44** [reply]
Do you know how a CDN would behave in case of a condition like the above (so for example based on User-Agent, it should redirect). Does/Will Vercel handle something like this out-of-the-box?

**Reply by ijjk** [reply]
Yeah Vercel already handles this out of the box

**Reply by balazsorban44** [reply]
I really appreciate your answers! :heart: 

Sorry, just to really clarify:

1. I am on an old browser, I visit /foo
2. /foo visited from old browser, redirect to /outdated
3. since it's redirect, /foo isn't cached in CDN
4. modern browser, I visit /foo
5. it gets served, and cached in CDN
6. Old browser, visits /foo.
7. New browser, visits /foo

Does the CDN somehow know that it needs to redirect on 6? Can it serve different cache results for different User-Agents?

Does 7 get a cached version or goes to origin?

(Is this what the Vary header https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary could be useful for if a CDN supports it?)
